### PR TITLE
Improve Mixin

### PR DESCRIPTION
**Pull request by LysandreJik**
Improves the Mixin in the hub by adding support for `api_endpoint`, `use_auth_token`, as well as `git_user` and `git_email`. 

It updates the tests so that they run in the staging environment. In doing so, I found a bug relative to absolute paths in `from_pretrained`, this fixes that at the same time.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -20,9 +20,13 @@ def get_version() -> str:
...
 extras = {}
 
+extras["torch"] = [
+    "torch",
+]
+
 extras["testing"] = [
     "pytest",
-]
+] + extras["torch"]
```
Comment (julien-c) [inline setup.py:29]: is there value in having a second CI setup with *no torch*? (to make sure we don't introduce a dependency on torch inadvertently down the road)

File: src/huggingface_hub/hub_mixin.py [hunk 2]
```diff
@@ -223,24 +227,47 @@ def push_to_hub(
...
 
-        token = HfFolder.get_token()
+        # The auth token is necessary to create a repo
+        if isinstance(use_auth_token, str):
+            huggingface_token = use_auth_token
+        elif use_auth_token is None and repo_url is not None:
+            # If the repo url exists, then no need for a token
+            huggingface_token = None
+        else:
+            huggingface_token = HfFolder.get_token()
```
Comment (julien-c) [inline src/huggingface_hub/hub_mixin.py:253]:
```suggestion
        if isinstance(use_auth_token, str):
            huggingface_token = use_auth_token
        elif use_auth_token:
            huggingface_token = HfFolder.get_token()
        else:
            huggingface_token = None
```

let's use the same logic as elsewhere no?

File: tests/test_hubmixin.py [hunk 3]
```diff
@@ -37,30 +46,83 @@ def forward(self, x):
...
+        self.assertEqual(len(files), 1)
+
         model.save_pretrained(
-            DUMMY_REPO_NAME, config={"num": 12, "act": "gelu"}, push_to_hub=True
+            f"{WORKING_REPO_DIR}/{REPO_NAME}", config={"num": 12, "act": "gelu"}
         )
+        files = os.listdir(f"{WORKING_REPO_DIR}/{REPO_NAME}")
+        self.assertTrue("config.json" in files)
+        self.assertTrue("pytorch_model.bin" in files)
+        self.assertEqual(len(files), 2)
```
Comment (julien-c) [inline tests/test_hubmixin.py:81]: nice

File: tests/test_hubmixin.py [hunk 4]
```diff
@@ -37,30 +46,83 @@ def forward(self, x):
...
+            f"tests/{WORKING_REPO_SUBDIR}/FROM_PRETRAINED",
+            config={"num": 10, "act": "gelu_fast"},
         )
-        model.save_pretrained(
-            "dummy-wts", config=None, push_to_hub=True, model_id=DUMMY_REPO_NAME
+
+        model = DummyModel.from_pretrained(
+            f"tests/{WORKING_REPO_SUBDIR}/FROM_PRETRAINED"
         )
+        self.assertTrue(model.config == {"num": 10, "act": "gelu_fast"})
```
Comment (julien-c) [inline tests/test_hubmixin.py:93]: `self.assertEqual` should work, no?

**Review by julien-c — COMMENTED** [review]
LGTM

File: setup.py (hunk 1 above)
Comment (LysandreJik) [reply setup.py:29]: Yes, I can add that

File: src/huggingface_hub/hub_mixin.py (hunk 2 above)
Comment (LysandreJik) [reply src/huggingface_hub/hub_mixin.py:253]: The issue with using the same logic as elsewhere is that the API becomes weird to use, in order to cover a rare use-case. Namely, the following will never work:

```py
model.push_to_hub("xxx")
```
it needs to be the following:
```py
model.push_to_hub("xxx", use_auth_token=True)
```
I fail to see when a user would want to push to the hub without having an `auth_token`, as it's necessary to create a repo. If the repo already exists and one wants to push to it, then the user already has to specify the `repo_url` to push to.

I think the API is cleaner by having `model.push_to_hub("xxx")` work if you already have an authentication token in your HF folder.

File: tests/test_hubmixin.py (hunk 4 above)
Comment (LysandreJik) [reply tests/test_hubmixin.py:93]: Indeed!

File: src/huggingface_hub/hub_mixin.py (hunk 2 above)
Comment (julien-c) [reply src/huggingface_hub/hub_mixin.py:253]: Ok

**Pull request merged**